src/runtime/matlab.cpp

```diff
@@ -0,0 +1,377 @@
+#include "HalideRuntime.h"
+
+#define INLINE inline __attribute__((always_inline))
+
+#ifndef MX_API_VER
+#define MX_API_VER 0x07040000
+#endif
+
+struct mxArray;
+
+// It is important to have the mex function pointer definitions in a
+// namespace to avoid silently conflicting symbols with matlab at
+// runtime.
+namespace Halide {
+namespace Runtime {
+namespace mex {
+
+enum { TMW_NAME_LENGTH_MAX = 64 };
+enum { mxMAXNAM = TMW_NAME_LENGTH_MAX };
+
+typedef bool mxLogical;
+typedef int16_t mxChar;
+
+enum mxClassID {
+    mxUNKNOWN_CLASS = 0,
+    mxCELL_CLASS,
+    mxSTRUCT_CLASS,
+    mxLOGICAL_CLASS,
+    mxCHAR_CLASS,
+    mxVOID_CLASS,
+    mxDOUBLE_CLASS,
+    mxSINGLE_CLASS,
+    mxINT8_CLASS,
+    mxUINT8_CLASS,
+    mxINT16_CLASS,
+    mxUINT16_CLASS,
+    mxINT32_CLASS,
+    mxUINT32_CLASS,
+    mxINT64_CLASS,
+    mxUINT64_CLASS,
+    mxFUNCTION_CLASS,
+    mxOPAQUE_CLASS,
+    mxOBJECT_CLASS,
+#if defined(_LP64) || defined(_WIN64)
+    mxINDEX_CLASS = mxUINT64_CLASS,
+#else
+    mxINDEX_CLASS = mxUINT32_CLASS,
+#endif
+
+    mxSPARSE_CLASS = mxVOID_CLASS
+};
+
+enum mxComplexity {
+    mxREAL,
+    mxCOMPLEX
+};
+
+#ifdef BITS_32
+typedef int mwSize;
+typedef int mwIndex;
+typedef int mwSignedIndex;
+#else
+typedef size_t mwSize;
+typedef size_t mwIndex;
+typedef ptrdiff_t mwSignedIndex;
+#endif
+
+typedef void (*mex_exit_fn)(void);
+typedef void (*mxFunctionPtr)(int, mxArray **, int, const mxArray **);
+
+// Given a halide type code and bit width, find the equivalent matlab class ID.
+WEAK mxClassID get_class_id(int32_t type_code, int32_t type_bits) {
+    switch (type_code) {
+    case halide_type_int:
+        switch (type_bits) {
```

Comment:
>  I'm not seeing the difference

because I am apparently suffering a brain lesion about which codebase I was reading, nevermind

> FWIW, emacs's

well there's your problem right there :-)